sound/soc/sof/mediatek/mt8195/mt8195.c

```diff
@@ -0,0 +1,360 @@
+		return -EINVAL;
+	}
+
+	if (adsp->dramsize < TOTAL_SIZE_SHARED_DRAM_FROM_TAIL) {
+		dev_err(dev, "adsp memory(0x%x) is not enough for share\n",
+			(u32)adsp->dramsize);
+		return -EINVAL;
+	}
+
+	dev_dbg(dev, "[ADSP] dram pbase=%pa, dramsize=0x%x\n",
+		&adsp->pa_dram, adsp->dramsize);
+
+	/* Parse CFG base */
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (!res) {
+		dev_err(dev, "no ADSP-CFG register resource\n");
+		return -ENXIO;
+	}
+	/* remap for DSP register accessing */
+	adsp->va_cfgreg = devm_ioremap_resource(dev, res);
+	if (IS_ERR(adsp->va_cfgreg))
+		return PTR_ERR(adsp->va_cfgreg);
+
+	adsp->pa_cfgreg = (phys_addr_t)res->start;
+	adsp->cfgregsize = (u32)resource_size(res);
+
+	dev_dbg(dev, "[ADSP] cfgreg-vbase=%p, cfgregsize=0x%x\n",
+		adsp->va_cfgreg, adsp->cfgregsize);
+
+	/* Parse SRAM */
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 1);
+	if (!res) {
+		dev_err(dev, "no SRAM resource\n");
+		return -ENXIO;
+	}
+
+	adsp->pa_sram = (phys_addr_t)res->start;
+	adsp->srammsize = (u32)resource_size(res);
+	if (adsp->srammsize < TOTAL_SIZE_SHARED_SRAM_FROM_TAIL) {
+		dev_err(dev, "adsp SRAM(0x%x) is not enough for share\n",
+			(u32)adsp->srammsize);
+		return -EINVAL;
+	}
+
+	dev_dbg(dev, "[ADSP] sram pbase=%pa,0x%x\n", &adsp->pa_sram, adsp->srammsize);
+
+	/* Parse MBOX base */
+	for (i = 0; i < DSP_MBOX_NUM; i++) {
+		res = platform_get_resource(pdev, IORESOURCE_MEM, 2 + i);
+		if (!res) {
+			dev_err(dev, "no MBOX resource %d\n", i);
+			return -ENXIO;
+		}
+
+		adsp->pa_mboxreg[i] = (phys_addr_t)res->start;
+		adsp->va_mboxreg[i] = devm_ioremap_resource(dev, res);
+		if (IS_ERR(adsp->va_mboxreg[i]))
+			return PTR_ERR(adsp->va_mboxreg[i]);
+
+		dev_dbg(dev, "[ADSP]MBOX%d pa_mboxreg:%pa, va_mboxreg:%p\n",
+			i, &adsp->pa_mboxreg[i], adsp->va_mboxreg[i]);
+	}
+
+	return ret;
+}
+
+static int adsp_sram_power_on(struct device *dev, bool val)
+{
+	void __iomem *va_dspsysreg;
+
+	va_dspsysreg = devm_ioremap(dev, ADSP_SRAM_POOL_CON, 0x4);
+	if (!va_dspsysreg) {
+		dev_err(dev, "error: failed to ioremap sram pool base 0x%x\n",
+			ADSP_SRAM_POOL_CON);
+		return -ENOMEM;
+	}
+
+	if (val)
+		writel(readl(va_dspsysreg) & ~DSP_SRAM_POOL_PD_MASK, va_dspsysreg);
+	else
+		writel(readl(va_dspsysreg) | DSP_SRAM_POOL_PD_MASK, va_dspsysreg);
+
+	return 0;
+}
+
+/*  Init the basic DSP DRAM address */
+static int adsp_memory_remap_init(struct device *dev)
+{
+	void __iomem *vaddr_emi_map;
+	struct adsp_chip_info *adsp;
+	int offset;
+
+	adsp = get_adsp_chip_data();
+	if (!adsp)
+		return -ENXIO;
+
+	vaddr_emi_map = devm_ioremap(dev, DSP_EMI_MAP_ADDR, 0x4);
+	if (!vaddr_emi_map) {
+		dev_err(dev, "error: failed to ioremap emi map base 0x%x\n",
+			DSP_EMI_MAP_ADDR);
+		return -ENOMEM;
+	}
+
+	offset = adsp->pa_dram - DRAM_PHYS_BASE_FROM_DSP_VIEW;
+	adsp->dram_offset = offset;
+	offset >>= DRAM_REMAP_SHIFT;
+	dev_dbg(dev, "adsp->pa_dram %llx, offset 0x%x\n", adsp->pa_dram, offset);
+	WARN_ON(offset < 0);
```

Comment:
first of all, it would be better to check for this apparent error immediately after calculating `offset` or maybe even before that by chacking `adsp->pa_dram < DRAM_PHYS_BASE_FROM_DSP_VIEW` perhaps where it's obtained from `res->start`. Secondly what does a `WARN_ON()` bring? Is this a valid value? If not, perhaps it shouldn't be used.